### ISSUE TITLE
[clang] -Wunreachable-code is unexpectedly triggered by expressions in the `noexcept` operator

### ISSUE THREAD

**Issue by frederick-vs-ja** [llvm/llvm-project#151501]
Code ([Godbolt link](https://godbolt.org/z/ja195jY13)):

```C++
template<int I, class T>
void get(T&&) {}

void test() {
  int n{};
  [[maybe_unused]] constexpr bool V = noexcept(::get<0>(n)) && noexcept(::get<1>(n));
}
```

Warning message:
```
<source>:6:73: warning: code will never be executed [-Wunreachable-code]
    6 |   [[maybe_unused]] constexpr bool V = noexcept(::get<0>(n)) && noexcept(::get<1>(n));
      |                                                                         ^~~~~~~~
1 warning generated.
Compiler returned: 0
```

This doesn't seem to be regression.

I _think_ this is false-positive as the operands of `noexcept` are always unevaluated, and we may want to calculate later `noexcept`-expressions for convenience.

**Comment by AaronBallman** [comment]
The first `noexcept` will always return `false` and so the `&&` short-circuits. When the first one isn't a short-circuit, the issue goes away: https://godbolt.org/z/hG4xazMnE But that behavior is specific to `noexcept`, you don't get the same out of `sizeof`: https://godbolt.org/z/zc8Ks16Tf

So  yeah, I think this is a false positive.
